File: count_sic_freq.pl
#!C:\Strawberry_Perl\bin\Perl -w

($pre) = @ARGV;

if ($pre eq 'Sector') {
	$total = 10;
} else {
	$total = 100;
}

for ($i = 1; $i <= $total; $i++) {
	$file = "$pre$i-company-list.txt";
	print "now processing $file\n";
	open(FI, $file);

	$count = 0;
	for ($j = 1; $j < 10000; $j++) {
		$freq{$j} = 0;
	}

	foreach $key (keys %freq) {
		if ($freq{$key} > 0) {
			$freq{$key} = 0;
		}
	}


	while (<FI>) {
		chomp;
		@line = split /,/;
		$company_name = $line[0];
#		$company_name =~ s/\"//g;
		$sic = $line[$#line];
		if ( ($sic > 0) && ($sic < 10000) ) {
			$freq{$sic}++;
			$count++;
		}
		else {
#			print "$company_name, $sic\n";
		}

	}
	close FI;
	print "count: $count\n";
	
#	foreach $key (keys %freq) {
#		if ($freq{$key} > 0) {
#			print "$key $freq{$key}\n";
#		}
#	}
	
	$f_file = "$pre$i-sic-freq.txt";
	open FIW, '>', $f_file;
	print FIW "#SIC #Occurrence #Percentage #Accumulated_Percentage\n";
	
	$accumu_perc = 0;
	foreach $key (sort { $freq{$b} <=> $freq{$a} } keys %freq) {
		if ($freq{$key} == 0) {
			last;
		}
		$percentage = $freq{$key} / $count;
		$accumu_perc += $percentage;
		print FIW "$key $freq{$key} $percentage $accumu_perc\n";
	}
	close FIW;
}

File: naming-evaluation/count_sic_freq.py
import sys

if __name__ == '__main__':
    
    prefix = sys.argv[1]

    totals = {'Industry': 100,
              'Sector': 10}
    
    for index in range(1, totals[prefix] + 1):
        
        print ("Index {index}".format(index=index))
        counts = {}
        
        with open('{prefix}{index}-company-list.txt'.format(
                    prefix=prefix, index=index),
                mode='r') as listfile:
            for line in listfile:
                sic = int(line.split()[-1])

                if sic:
                    counts[sic] = counts.get(sic, 0) + 1
        
        siclist = list((count, sic) for sic, count in counts.items())
        siclist.sort(reverse=True)
        
        totalcount = sum(count for count, sic in siclist)
        
        with open('{prefix}{index}-sic-freq.txt'.format(prefix=prefix,
                                                        index=index),
                                                        mode='w') as f:
            f.write('#SIC #Occurrence #Percentage #Accumulated_Percentage\n')
            accumu_prec = 0            
            for count, sic in siclist:
                percentage = count/totalcount
                accumu_prec += percentage  
                f.write('{sic} {freq} {percentage} {accumu_prec}\n'.format(
                    sic=sic, freq=count, percentage=percentage,
                    accumu_prec=accumu_prec))
